src/runtime.rs

```diff
@@ -137,36 +147,95 @@ impl Runtime {
     pub fn new() -> Self {
         Self {
             spawner: Spawner(Rc::new(JunkSpawner)),
-            revision: Revision(0),
             cache: SharedLocalCache::default(),
-            wk: noop_waker(),
+            rcs: Arc::new(RwLock::new(RevisionControlSystem {
+                revision: Revision(0),
+                waker: noop_waker(),
+                // require the first revision to be forced?
+                pending_changes: AtomicBool::new(false),
+            })),
         }
     }
 
-    /// The current revision of the runtime, or how many times `run_once` has
-    /// been invoked.
+    /// The current revision of the runtime, or how many runs occurred.
     pub fn revision(&self) -> Revision {
-        self.revision
+        self.rcs.read().revision
     }
 
-    /// Runs the root closure once with access to the runtime context,
-    /// increments the runtime's `Revision`, and drops any cached values
-    /// which were not marked alive.
+    /// Runs the root closure once with access to the runtime context and drops
+    /// any cached values which were not marked alive. `Revision` is
+    /// incremented at the start of a run.
     pub fn run_once<Out>(&mut self, op: impl FnOnce() -> Out) -> Out {
-        self.revision.0 += 1;
+        self.execute(op, self.rcs.write())
+    }
+
+    /// Runs the root closure once with access to the runtime context and drops
+    /// any cached values which were not marked alive. `Waker` is set for the
+    /// next `Revision`, which starts after the start of the run.
+    pub fn run_once_with<Out>(&mut self, op: impl FnOnce() -> Out, waker: Waker) -> Out {
+        let mut rcs_write = self.rcs.write();
+        rcs_write.waker = waker;
+        self.execute(op, rcs_write)
+    }
+
+    /// Forces the next `Revision` without any changes.
+    pub fn force(&self) {
+        self.rcs.read().pending_changes.store(true, std::sync::atomic::Ordering::Relaxed);
+    }
+
+    /// If change occured durig the last `Revision` then calls `run_once`
+    /// else returns `Poll::Pending`. It is required to force your
+    /// first revision to register your state variables!
+    pub fn poll_once<Out>(&mut self, op: impl FnOnce() -> Out) -> Poll<Out> {
+        // Avoid write lock
+        let rcs_read = self.rcs.upgradable_read();
+        if !rcs_read.pending_changes.load(std::sync::atomic::Ordering::Relaxed) {
```

Comment:
With my suggestions above, you can make this a method that doesn't require taking even a read lock, I'd suggest putting this check in a method like `RevisionController::has_pending_changes`.